src/Microsoft.ApiDesignGuidelines.Analyzers/Core/ImplementIDisposableCorrectly.cs

```diff
@@ -106,13 +112,494 @@ public abstract class ImplementIDisposableCorrectlyAnalyzer : DiagnosticAnalyzer
+                    if (suppressFinalizeMethod == null)
+                    {
+                        return;
+                    }
+
+                    var analyzer = new Analyzer(context.Compilation, disposableType, disposeInterfaceMethod, suppressFinalizeMethod);
+                    analyzer.Initialize(context);
+                });
+        }
+
+        private static bool IsDisposeBoolMethod(IMethodSymbol method)
+        {
+            if (method.Name == DisposeMethodName && method.MethodKind == MethodKind.Ordinary &&
+                method.ReturnsVoid && method.Parameters.Length == 1)
+            {
+                var parameter = method.Parameters[0];
+                if (parameter.Type != null && parameter.Type.SpecialType == SpecialType.System_Boolean && parameter.RefKind == RefKind.None)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Analyzes single instance of compilation.
+        /// </summary>
+        private class Analyzer
+        {
+            private readonly Compilation compilation;
+            private readonly INamedTypeSymbol disposableType;
+            private readonly IMethodSymbol disposeInterfaceMethod;
+            private readonly IMethodSymbol suppressFinalizeMethod;
+
+            public Analyzer(Compilation compilation, INamedTypeSymbol disposableType, IMethodSymbol disposeInterfaceMethod, IMethodSymbol suppressFinalizeMethod)
+            {
+                this.compilation = compilation;
+                this.disposableType = disposableType;
+                this.disposeInterfaceMethod = disposeInterfaceMethod;
+                this.suppressFinalizeMethod = suppressFinalizeMethod;
+            }
+
+            public void Initialize(CompilationStartAnalysisContext context)
+            {
+                context.RegisterSymbolAction(AnalyzeNamedTypeSymbol, SymbolKind.NamedType);
+                context.RegisterOperationBlockAction(AnalyzeOperationBlock);
+            }
+
+            private void AnalyzeNamedTypeSymbol(SymbolAnalysisContext context)
+            {
+                var type = context.Symbol as INamedTypeSymbol;
+                if (type != null && type.IsType && type.TypeKind == TypeKind.Class)
+                {
+                    var implementsDisposableInBaseType = ImplementsDisposableInBaseType(type);
+
+                    if (ImplementsDisposableDirectly(type))
+                    {
+                        var disposeMethod = FindDisposeMethod(type);
+                        if (disposeMethod != null)
+                        {
+                            // This is difference from FxCop implementation
+                            // IDisposable Reimplementation Rule is violated only if type re-implements Dispose method, not just interface
```

Comment:
Interesting. Normally we try to keep the rule functionality the same as FxCop in order to not flood users with new warnings and to make testing easier (we can compare with fxcop results).

In this case, however, it looks like we are being more strict and therefore should produce fewer false-positives than FxCop - is that right? Was that the motivation for this deviation from FxCop?

Fewer false positives sounds good to me. @lgolding - what do you think?